rfcs/20231012-external-mel-bindings.md

```diff
@@ -0,0 +1,546 @@
+- RFC PR: tbd
+
+# Summary
+
+This RFC proposes a new approach to writing bindings to JavaScript functions in
+Melange. 
+
+# Basic example
+
+The proposal consists on introducing three extensions:
+- `let%js` for declaration of bindings to JavaScript plain values / functions
+- `let%js.import` for declaration of bindings to JavaScript modules
+- `let%js.obj` would work essentially like `mel.obj`
+
+## `let.js`
+
+This extension operates over `let` bindings, and are expected to assign a string
+literal to a value. The string literal will be interpreted as plain JavaScript,
+and will define the operations to apply over the labelled arguments, which can
+be referenced inside the string by using the `$` prefix.
+
+For example, a binding to `decodeURI` global function:
+
+```ocaml
+let%js decodeURI : uri:string -> string = {|decodeURI($uri)|}
+```
+
+Note that the usage of quoted string `{||}` is optional, but recommended, as
+double quotes `"` inside the string don't need to be escaped.
+
+Also note the choice of `let` vs `external` is intentional: besides having to
+type less characters, quoted strings remain untouched by OCamlformat, while they
+get transformed into regular strings in externals.
+
+The syntactic construct chosen is also consistent with the generated code:
+unlike `external` declarations, `let` bindings don't get inlined at the call
+point.
+
+If the author of the binding doesn't want to use labeled arguments, they can
+append an underscore to have them removed from the generated OCaml function:
+
+```ocaml
+let%js decodeURI : uri_:string -> string = "decodeURI($uri_)"
+let _ : string = decodeURI "foo" (* no warning about a missing label ever *)
+```
+
+The preprocessor of `let%js` will check that type definitions to have at least
+an arrow type. So in cases where before one would just have some abstract type
+`t` in the external, now they would become `unit -> t`.
+
+For example, the current external to `Math.PI`:
+
+```ocaml
+external pi : float = "PI" [@@mel.scope "Math"]
+```
+
+would be written as:
+
+```ocaml
+let%js pi : float = "Math.PI"
+```
+
+## `let%js.import`
+
+The other extension introduced, `let%js.import`, is used for importing values
+from JavaScript modules into the current OCaml module:
+
+```ocaml
+type t
+let%js.import foo: t = "./bar"
+```
+
+Gets transformed into this JavaScript:
+
+```js
+import foo from "./bar";
+```
+
+It can leverage the OCaml destructuring feature to pick values from the imported
+module, but all of them need type annotations:
+
+
+```ocaml
+type u
+let%js.import {foo = (hey : string); bar: int} = "./bar"
+```
+Gets transformed into this JavaScript:
+
+```js
+import {foo as hey, bar} from "./bar";
+```
+
+Note there is no need to predefine a record type before using `let%js.import`,
+as the record is used merely syntactically to be transformed into an `import`
+statement.
+
+## `let%js.obj`
+
+Would be similar to `mel.obj` but as an extension applied to `let` bindings:
```

Comment:
so it's never possible to construct a JS object without using a `let` binding?

how would you express `Js.log([%mel.obj { foo= "bar"}])`